modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/distributed/near/IgniteCacheDistributedQueryDefaultTimeoutSelfTest.java

```diff
@@ -0,0 +1,196 @@
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_1, 500, TimeUnit.MILLISECONDS, true, true);
+    }
+
+    /** */
+    @Test
+    public void testRemoteQueryWithMergeTableTimeout() throws Exception {
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_2, 500, TimeUnit.MILLISECONDS, true, false);
+    }
+
+    /** */
+    @Test
+    public void testRemoteQueryExecutionCancel0() throws Exception {
+        testQueryCancel(CACHE_SIZE, VAL_SIZE, QRY_1, 1, TimeUnit.MILLISECONDS, false, true);
+    }
+
+    /** */
+    private void testQueryCancel(int keyCnt, int valSize, String sql, int timeoutUnits, TimeUnit timeUnit,
+        boolean timeout, boolean checkCanceled) throws Exception {
+        try (Ignite client = startGrid("client")) {
+            IgniteCache<Object, Object> cache = client.cache(DEFAULT_CACHE_NAME);
+
+            assertEquals(0, cache.localSize());
+
+            int p = 1;
+            for (int i = 1; i <= keyCnt; i++) {
+                char[] tmp = new char[valSize];
+                Arrays.fill(tmp, ' ');
+                cache.put(i, new String(tmp));
+
+                if (i / (float)keyCnt >= p / 10f) {
+                    log().info("Loaded " + i + " of " + keyCnt);
+
+                    p++;
+                }
+            }
+
+            assertEquals(0, cache.localSize());
+
+            SqlFieldsQuery qry = new SqlFieldsQuery(sql);
+
+            final QueryCursor<List<?>> cursor;
+            if (timeout) {
+                qry.setTimeout(timeoutUnits, timeUnit);
+
+                cursor = cache.query(qry);
+            }
+            else {
+                cursor = cache.query(qry);
+
+                client.scheduler().runLocal(new Runnable() {
+                    @Override public void run() {
+                        cursor.close();
+                    }
+                }, timeoutUnits, timeUnit);
+            }
+
+            try (QueryCursor<List<?>> ignored = cursor) {
+                cursor.getAll();
+
+                if (checkCanceled)
+                    fail("Query not canceled");
+            }
+            catch (CacheException ex) {
+                error("Got expected exception", ex);
+
+                assertNotNull("Must throw correct exception", X.cause(ex, QueryCancelledException.class));
+            }
+
+            // Give some time to clean up.
+            Thread.sleep(TimeUnit.MILLISECONDS.convert(timeoutUnits, timeUnit) + 3_000);
+
+            checkCleanState();
+        }
+    }
+
+    /**
+     * Validates clean state on all participating nodes after query cancellation.
+     */
+    private void checkCleanState() {
+        for (int i = 0; i < GRIDS_CNT; i++) {
+            IgniteEx grid = grid(i);
+
+            // Validate everything was cleaned up.
+            ConcurrentMap<UUID, ?> map = U.field(((IgniteH2Indexing)U.field((GridProcessor)U.field(
+                grid.context(), "qryProc"), "idx")).mapQueryExecutor(), "qryRess");
```

Comment:
🤦‍♂️
qryProc is exposed via context.query() And idx is exposed via qryProc.getIndexing(). So the final code should be like this:
```suggestion
            ConcurrentMap<UUID, ?> map = U.field(((IgniteH2Indexing)grid().context().query()
            .getIndexing()).mapQueryExecutor(), "qryRess");
```